Revert endorsement and reference triple records to original

nedmsmith opened the pull request:
This PR reverts `endorsement-triple-record` and `reference-triple-record` to original CDDL by adding array brackets around `measurement-map` to allow multiplicity of measurements for a given environment.

The Measurements and Measurement Keys sections are updated to better describe measurement multiplicity and disambiguation of measurement values.

thomas-fossati commented on draft-ietf-rats-corim.md:
```diff
@@ -790,25 +791,29 @@ The supply chain entity that is responsible for providing the the measurements (
 is by default the CoRIM signer. If a different entity is authorized to provide measurement values,
 the `authorized-by` statement can be supplied in the `measurement-map`.
 
+An environment may have multiple measurements of the same type.
+Measurement keys may be used to disambiguate these measurements.
```
WFM, but another way to put this (which I find more intuitive) is to say that an environment can be composed of multiple individually measured "elements."

I think "element" is the piece of vocabulary we were missing before Dionna proposed it in #257, allowing speaking about multiplicity more naturally.

Quoting from #257:
```
An environment has one or more measured elements.
Measured elements are distinguished from each other by the `measurement-map` key.
A measured element can have a dedicated measurement, such as a security version number.
Multiple elements could be combined into a single measurement, such as a running hash of multiple files.
A measured element can also have several measurement values, such as a name, version, and digest.
Measurements can have class, instance or group scope.
A `measurement-map`'s scope is determined by the `environment-map` it is paired with.
```

thomas-fossati commented on draft-ietf-rats-corim.md:
```diff
@@ -790,25 +791,29 @@ The supply chain entity that is responsible for providing the the measurements (
 ###### Measurement Keys {#sec-comid-mkey}
 
-The types defined for a measurement identifier are OID, UUID or uint.
+Measurement keys are locally scoped identifiers that disambiguate multiple instances of measurements of the same type.
```
It could be that, but it could also be they serve as "element" identifiers.

yogeshbdeshpande replied:
Agree with Thomas, in the latest commit, I have added this aspect, please have a look!

deeglaze commented:
The matching semantics wording needs to be updated to say that all measurement-maps have to match the ACS for the triple to apply.

nedmsmith commented:
I modified Section 8.1 which defines the ECT structure, to reflect that claims-map always has multiplicity and to tweak naming to better reflect the idea that measurements can have "element IDs".

I also modified the corroboration section (sorry Thomas) to better reflect the matching logic in terms of the internal representation. The before text was blending internal and external representation naming.

nedmsmith commented on draft-ietf-rats-corim.md:
```diff
@@ -1595,17 +1602,18 @@ Profile (label 6):
   c: claims-map
 }
-local-environment =  bstr / tstr
+element-id =  bstr / tstr / uint / tagged-oid-type / tagged-uuid-type
```
Note: element-id could go back to being bstr / tstr if the mapping from external to internal describes a mapping from uint, tagged-oid-type and tagged-uuid-type to bstr.

deeglaze commented on draft-ietf-rats-corim.md:
```diff
@@ -1932,12 +1940,13 @@ This can be acheived by sorting the triples before processing, by repeating proc
+The Verifier compares each `rv` reference value condition ECT against ECT entries in the ACS as described in {{sec-match-one-se}}.
 
-If any `reference-triple-record` in the Reference Value triple does not match the ACS then the entire triple is ignored.
+Foreach entry in the `rv` list, if any part of an `rv` condition does not match an ACS entry, then the entire `rv` additon is not added to the ACS.
```
```suggestion
For each entry in the `rv` list, if any part of an `rv` condition does not match an ACS entry, then the entire `rv` addition is not added to the ACS.
```

thomas-fossati commented on draft-ietf-rats-corim.md:
```diff
@@ -790,25 +791,31 @@ The supply chain entity that is responsible for providing the the measurements (
+Measurement keys are locally scoped identifiers.
+`Mkey` values may be necessary to disambiguate multiple measurements of the same type.
+
+`Mkey` identifiers can be either OID, UUID or uint.
```
`mkey` is an extensible type.  This aspect should be highlighted.

yogeshbdeshpande replied:
Have added a line as suggested by Thomas!

thomas-fossati commented on draft-ietf-rats-corim.md:
```diff
@@ -1595,17 +1602,18 @@ Profile (label 6):
   c: claims-map
 }
-local-environment =  bstr / tstr
+element-id =  bstr / tstr / uint / tagged-oid-type / tagged-uuid-type
```
the typographical convention for closed type choices is `NAME-type-choice`, so this should look like:

```
element-id-type-choice /= bstr
element-id-type-choice /= tstr
element-id-type-choice /= uint
element-id-type-choice /= tagged-oid-type
element-id-type-choice /= tagged-uuid-type
```

thomas-fossati replied:
Clarifying question: are you suggesting avoiding using the typographical conventions in S3 for CDDL related to internal representation?
If so, I think it'd be better to keep the conventions and use an ad hoc prefix/suffix/infix for internal, e.g.: `__`, like in good old C :-)

nedmsmith replied:
We agreed to define the internal representation of element ID to match the external representation and to add wording to the effect that element IDs are compared in a byte for byte approach.

nedmsmith replied:
Fixed in commit: https://github.com/ietf-rats-wg/draft-ietf-rats-corim/commit/0d82b15fffd8cef5d09f4747e5e45bed8dd1caf6

nedmsmith commented:
See issue #266

yogeshbdeshpande commented:
We still need to work on other triples involving `measurement-values-map`

yogeshbdeshpande commented on draft-ietf-rats-corim.md:
```diff
@@ -1595,17 +1603,18 @@ Profile (label 6):
 ECT = {
   ? e: environment-map
-  ? c: claims-map / [ + local-claim ]
+  ? c: [ + local-claim ]
```
Line 1583 which describes the claims needs to be:
1. Reworded with a more appropriate name, like a set of claims 
2. Needs to explain better, regarding the claims pertaining to each of the measured elements belonging to a target environment

nedmsmith replied:
Fixed in commit: https://github.com/ietf-rats-wg/draft-ietf-rats-corim/pull/268/commits/59fd80d9894bf52b9a6718d9339d6548c238ee9f

yogeshbdeshpande commented on draft-ietf-rats-corim.md:
```diff
@@ -790,25 +791,32 @@ The supply chain entity that is responsible for providing the the measurements (
+`Mkey` may be necessary to disambiguate multiple measurements of the same type.
+Alternatively, `Mkey` may also be used to  distinguish multiple measured elements in an environment.
+
+`Mkey` identifiers can be either OID, UUID or uint. `Mkey` is an extensible type.
```
```suggestion
`Mkey` is an extensible identifier. The initial types defined are OID, UUID, and uint.
```

henkbirkholz replied:
```suggestion
`mkey` identifiers can be either OID, UUID or uint. `Mkey` is an extensible type.
```

henkbirkholz commented on draft-ietf-rats-corim.md:
```diff
@@ -790,25 +791,32 @@ The supply chain entity that is responsible for providing the the measurements (
 
-The types defined for a measurement identifier are OID, UUID or uint.
+Measurement keys are locally scoped identifiers.
+`Mkey` may be necessary to disambiguate multiple measurements of the same type.
```
```suggestion
`mkey` may be necessary to disambiguate multiple measurements of the same type.
```

henkbirkholz commented on draft-ietf-rats-corim.md:
```diff
@@ -790,25 +791,32 @@ The supply chain entity that is responsible for providing the the measurements (
+`mkey` may be necessary to disambiguate multiple measurements of the same type.
+Alternatively, `Mkey` may also be used to  distinguish multiple measured elements in an environment.
+
+`Mkey` is an extensible identifier. The initial types defined are OID, UUID, and uint.
```
```suggestion
`mkey` is an extensible identifier. The initial types defined are OID, UUID, and uint.
```

deeglaze commented on draft-ietf-rats-corim.md:
```diff
@@ -1932,12 +1941,13 @@ This can be acheived by sorting the triples before processing, by repeating proc
 
-The Verifier compares each `reference-triple-record` against ACS entries as described in {{sec-match-one-se}}, where the `reference-triple-record` takes the place of a `stateful-environment-record`.
-If all fields of the `reference-triple-record` match the ACS, then the Verifier MUST add the RVP authority to each matching ACS field.
+The Verifier compares each `rv` reference value condition ECT against ECT entries in the ACS as described in {{sec-match-one-se}}.
```
The referenced section is not properly named for what this is needed to match all mkey/mval pairs

"Processing a stateful environment against the Appraisal Claims Set {#sec-match-one-se}"

nedmsmith replied:
The corroboration section removed the reference to sec-match-one-se for now.
The wording now says "... ECTs are identical...". There should be a section that describes how two ECTs are compared. Once that section is complete, a reference to it can be added here.

andrew-draper reviewed (changes requested):
This change is incomplete, it changes some external representation syntax, but not other representations.
In dicussions today we were unable to come to an agreement about whether the other triples should also be modified.
I think this PR should change all triples at the same time so we can have one set of explanatory text.

deeglaze reviewed (approved):
Issue#279 opened to close the matching semantics gap.

yogeshbdeshpande commented on draft-ietf-rats-corim.md:
```diff
@@ -1590,20 +1596,22 @@ Profile (label 6):
-  ? cm: cm-type
-  ? p: $profile-type-choice
+  ? environment: environment-map
+  ? properties: [ + properties-map ]
```
Given that `ECT` is an acronym for `Environment Claims Tuple`: I would rename line 1604 on the right to 

? iclaims: [+internal-claims-map]

&

Line 1610 to be:
internal-claims-map = {
? eid: element-id
claims: measurement-values-map
}

nedmsmith replied:
I keyed off of the existing prose that describes it as properties. I don't think we need to bow to the consistency gods by changing ECT to EPT since the main focus is still on claims. In other words, the measurement-values-map contents are still the "claims". 

I find it easier to write the prose if the structures are all the same but with slight modification. 

Given we now describe environments as consisting of zero or more elements. It may be appropriate to name `properties-map` to `element-map` then the contents of an element-map is the element-id and the claims `claims-map`.

yogeshbdeshpande replied:
Works for me and I admit a very sensible suggestion.

yogeshbdeshpande reviewed (commented):
Some editorial comments but in general LGTM!

yogeshbdeshpande reviewed (approved):
LGTM!

thomas-fossati commented on draft-ietf-rats-corim.md:
```diff
@@ -1552,58 +1558,38 @@ Attestation Results Set (ARS):
+Conceptual Messages are Verifier input and output values such as Evidence, Reference Values, Endorsed Values, Appraisal Policy, and Attestation Results.
 
-: Properties of the Target Environment.
+The internal representation of Conceptual Messages, as well as the ACS {{sec-ir-acs}} and ARS {{sec-ir-ars}}, are constructed from a common building block structure called Environment-Claims Tuple (ECT).
```
```suggestion
The internal representation of Conceptual Messages, as well as the ACS ({{sec-ir-acs}}) and ARS ({{sec-ir-ars}}), are constructed from a common building block structure called Environment-Claims Tuple (ECT).
```

thomas-fossati commented on draft-ietf-rats-corim.md:
```diff
@@ -1552,58 +1558,38 @@ Attestation Results Set (ARS):
 
-Name Space (label 4):
+2. Elements : Identifies the set of elements contained within a Target Environment and their trustworthiness Claims.
 
-: Identifies the name space from which the tuple was created.
+3. Authority : Identifies the entity that issued the tuple. A certain type of key material by which the authority (and corresponding provenance) of the tuple can be determined, such as the public key of an asymmetric key pair that is associated with an authority's PKIX certificate.
 
-CM Type (label 5):
+4. Conceptual Message Type : Identifies the type of Conceptual Message that originated the tuple.
 
-: Identifies the type of Conceptual Message that originated the tuple.
+5. Profile : The profile that defines this tuple. If no profile is used, this attribute is omitted.
```
Purely typographic comment: there are structural elements that can be used to assemble an enumeration/definition list, rather than doing it by hand.  E.g.:

```
{: vspace="0"}
Environment:
: Identifies ...

Elements:
: Identifies ...
```

or

```
{:my-enum: style="format %d."}

{: my-enum}
* Environment: Indicates ...
* Elements: Indicates ...
```

The advantage is that rendering in the supported output formats (TXT, HTML, PDF) is taken care of automatically - and, hopefully, also optimally.

nedmsmith replied:
Fixed in commit https://github.com/ietf-rats-wg/draft-ietf-rats-corim/pull/268/commits/7a01ca689f4013783ba5138c0d80232214e3976b

thomas-fossati commented on draft-ietf-rats-corim.md:
```diff
@@ -1633,9 +1618,9 @@ ae = [
 ~~~
 
-| Type | `e` | `c` | `a` | `ns` | `cm` | `p` |
+| Type | `en` | `el` | `a` | `cm` | `p` |
```
The shorthand convention doesn't seem to be defined anywhere.  OK, it's not completely unintuitive :-) but can't we use the unabbreviated labels?

thomas-fossati commented on draft-ietf-rats-corim.md:
```diff
@@ -1843,15 +1828,37 @@ The `cm` field is set based on the type of Conceptual Message inputted or to be
 [^issue] https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/96
 
-#### Reference and Endorsed Values Tranformation
+#### Reference Triples Tranformation {#sec-ref-trans}
```
```suggestion
#### Reference Triples Transformation {#sec-ref-trans}
```

thomas-fossati commented on draft-ietf-rats-corim.md:
```diff
@@ -1843,15 +1828,37 @@ The `cm` field is set based on the type of Conceptual Message inputted or to be
-#### Reference and Endorsed Values Tranformation
+#### Reference Triples Tranformation {#sec-ref-trans}
+
+An available `rv` list entry {{sec-ir-ref-val}} is allocated.
```
This (and similar instances below) will render as:
```
An available rv list entry Section 8.2.1.2 is allocated.
```
which is not great.

Please wrap all the references to sections in parentheses so that it reads like:
```
An available rv list entry (Section 8.2.1.2) is allocated.
```

thomas-fossati commented on draft-ietf-rats-corim.md:
```diff
@@ -1843,15 +1828,37 @@ The `cm` field is set based on the type of Conceptual Message inputted or to be
+
+#### Endorsement Triples Tranformations {#sec-end-trans}
+
+Endorsement Triple Transformation :
+
+: An available `ev` entry {{sec-ir-end-val}} is allocated.
+The Endorsement Triple {{sec-comid-triple-endval}} `environment-map` is copied to the `environment-map` for both the `ev` `condition` and `ev` `addition` ECTs.
+
+: For each `measurement-map` entry in the measurements list, the i<sup>th</sup> `measurement-map` entry is copied to the i<sup>th</sup> entry in the `addition` ECT `elements` list.
+
+: The issuer of the Endorsement conceptual message is copied to the `ev` `addition` ECT authority field.
+
+: If the Endorsement conceptual message has a profile, the profile is copied to the `ev` `addition` ECT profile field.
```
What is the typographical intent here?

If this is an algorithm, use enumerations (maybe custom ones, like I described in a comment above).

Anonymous definition lists doesn't look like the right tool for the job.

nedmsmith replied:
custom enumerations are new to me. I have to play around with it.

thomas-fossati replied:
You could do:

```
{:algo-steps: counter="bar" style="format Step %d."}

{:algo-steps}
* An available `rv` list entry {{sec-ir-ref-val}} is allocated.
The Reference Values Triple {{sec-comid-triple-refval}} `environment-map` is copied to the `environment-map` for both the `rv` `condition` and `rv` `addition` ECTs.
* For each `measurement-map` entry in the measurements list, the i<sup>th</sup> `measurement-map` is copied to the i<sup>th</sup> `element-map` in the `elements` list of the `rv` `addition` ECT.
* The issuer of the Endorsement conceptual message is copied to the `ev` `addition` ECT authority field.
* If the Endorsement conceptual message has a profile, the profile is copied to the `ev` `addition` ECT profile field.
```

which is rendered as:
```
   Step 1.  An available rv list entry Section 8.2.1.2 is allocated.
            The Reference Values Triple Section 5.1.4.2 environment-map
            is copied to the environment-map for both the rv condition
            and rv addition ECTs.

   Step 2.  For each measurement-map entry in the measurements list, the
            i^th measurement-map is copied to the i^th element-map in
            the elements list of the rv addition ECT.

   Step 3.  The issuer of the Endorsement conceptual message is copied
            to the ev addition ECT authority field.

   Step 4.  If the Endorsement conceptual message has a profile, the
            profile is copied to the ev addition ECT profile field.
```

thomas-fossati commented on draft-ietf-rats-corim.md:
```diff
@@ -1843,15 +1828,37 @@ The `cm` field is set based on the type of Conceptual Message inputted or to be
+
+: If the Endorsement conceptual message has a profile, the profile is copied to the `ev` `addition` ECT profile field.
+
+Conditional Endorsement Triple Transformation :
+
+Conditional Endorsement Series Triple Transformation :
```
Are these TODOs (in which case, we should mark them explicitly and link to their associated issues)?  Or did you just forget to add the content?

thomas-fossati replied:
OK; stick a `[^issue] https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/XYZ` nearby then.

thomas-fossati commented on draft-ietf-rats-corim.md:
```diff
@@ -1928,36 +1935,35 @@ This can be acheived by sorting the triples before processing, by repeating proc
-Reference Value Providers (RVP) publish Reference Values triples that are matched against ACS entries.
-Reference Values may describe multiple acceptable states for Attesters; hence "matching" determines that Evidence (contained in the ACS) satisfies an appropriate subset of the available Reference Values.
-If the appropriate subset matches, the authority of the RVP is added to the appropriate ACS entries.
+Reference Value Providers (RVP) publish Reference Values using the Reference Values Triple {{sec-comid-triple-refval}} which are transformed {{sec-ref-trans}} into an internal representation {{sec-ir-ref-val}}. Reference Values may describe multiple possible Attester states.
```
parentheses deficiency syndrome :-)

andrew-draper reviewed (approved):
I still think this is incomplete, but we have a direction for fixing the missing parts in future changes, so I think it is ok to check in this change.
